### DISCUSSION TITLE
Best practice:  update HA & reboot node red from inside container

### DISCUSSION THREAD

**Discussion by GeorgeCastanza** [tteck/Proxmox#1700] [Q&A] [answered]
I've been using Proxmox and these awesome helper scripts for a couple of months, and been doing upgrades (HA etc) from within the container.  Is this the best way?  I noticed here that someone asked a question about upgrading by running the helper script.  Not sure what is best practice??

Second, I run node red in proxmox.  Previously, running node red on bare metal, I shut node red down by **node-red-stop** on the command line.  That doesnt work in proxmox.  Is there a way to shut down node red inside the container before shutting it down on the proxmox screen (reboot, shutdown buttons on the pve page)   

Thanks for these scripts and help for newbs like me.

**Comment by tteck** [comment]
When available, you have the option to execute the script within the LXC console to perform updates or utilize other features.

Node-Red commands
```
systemctl start nodered
systemctl stop nodered
systemctl restart nodered
```

**Reply by tteck** [reply, marked as answer]
I make an effort to steer clear of running bash scripts from external sources that I cannot oversee within my own scripts. The alteration to the installation was implemented in this [commit](https://github.com/tteck/Proxmox/commit/283c328917e4ca80841b654e4df99aa1b436b690).

Should you favor the commands employed by the Node-RED Installer script, you can readily configure it by executing the subsequent commands within the Node-Red LXC console.

```
echo "journalctl -f -n 100 -u nodered -o cat" >/usr/bin/node-red-log
chmod +x /usr/bin/node-red-log
echo "systemctl stop nodered" >/usr/bin/node-red-stop
chmod +x /usr/bin/node-red-stop
echo "systemctl start nodered" >/usr/bin/node-red-start
chmod +x /usr/bin/node-red-start
echo "systemctl restart nodered" >/usr/bin/node-red-restart
chmod +x /usr/bin/node-red-restart
```

EDIT: I added the commands to the script.

**Comment by tteck** [comment]
Which variant of Home Assistant are you utilizing? Is it Home Assistant Core or Home Assistant Container?

**Reply by GeorgeCastanza** [reply]
Home Assistant Core.

**Reply by tteck** [reply]
You can apply the same approach employed in Node-Red for Home Assistant Core.

```
# ha-update
echo "systemctl stop homeassistant.service && source /srv/homeassistant/bin/activate && pip3 install --upgrade homeassistant && systemctl start homeassistant.service && exit" >/usr/bin/ha-update
chmod +x /usr/bin/ha-update

# ha-update-beta
echo "systemctl stop homeassistant.service && source /srv/homeassistant/bin/activate && pip3 install --pre --upgrade homeassistant && systemctl start homeassistant.service && exit" >/usr/bin/ha-update-beta
chmod +x /usr/bin/ha-update-beta

# ha-update-dev
echo "systemctl stop homeassistant.service && source /srv/homeassistant/bin/activate && pip3 install --upgrade git+https://github.com/home-assistant/core.git@dev && systemctl start homeassistant.service && exit" >/usr/bin/ha-update-dev
chmod +x /usr/bin/ha-update-dev

# ha-update-script
echo "bash -c \"\$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/homeassistant-core.sh)\"" >/usr/bin/ha-update-script
chmod +x /usr/bin/ha-update-script
```
This introduces four new commands: `ha-update` for updating to the stable branch, `ha-update-beta` for updating to the beta branch, `ha-update-dev` for updating to the dev branch, and `ha-update-script`, which triggers the script.

**Comment by tteck** [comment]
I'm considering the incorporation of an `update` command across all LXCs. This command would invoke the update script, simplifying the process by eliminating the need to locate, copy, and paste the bash script in the LXC console. Instead, you would only need to execute the `update` command.

**Reply by tteck** [reply]
The following applications now include the `update` command for testing purposes:
- ESPHome
- Home Assistant Core
- Home Assistant
- Node-Red
- Omada
- Zigbee2MQTT

To manually integrate the `update` command, modify `<app>` to match the precise application name, and then execute the command within the LXC console:
```
echo "bash -c \"\$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/<app>.sh)\"" >/usr/bin/update
chmod +x /usr/bin/update
```
